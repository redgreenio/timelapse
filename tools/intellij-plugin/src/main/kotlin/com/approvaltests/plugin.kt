package com.approvaltests

const val ID = "io.redgreen.timelapse.dev"
const val DISPLAY_NAME = "Approval Tests"
