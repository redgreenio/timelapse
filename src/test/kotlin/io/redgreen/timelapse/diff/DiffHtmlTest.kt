package io.redgreen.timelapse.diff

import org.approvaltests.Approvals
import org.junit.jupiter.api.Test

class DiffHtmlTest {
  @Test
  fun `single diff section`() {
    // given
    val diffWithSingleSection = """
      diff --git a/build.gradle b/build.gradle
      index 4604741..2bc696d 100644
      --- a/build.gradle
      +++ b/build.gradle
      @@ -1,6 +1,6 @@
       buildscript {
         apply from: 'buildscripts/plugins.gradle'
      -  ext.kotlin_version = '1.3.70'
      +  ext.kotlin_version = '1.3.72'
       
         repositories {
           mavenCentral()
    """.trimIndent()

    // when
    val diffHtml = FormattedDiff.from(diffWithSingleSection).toHtml()

    // then
    Approvals.verify(diffHtml)
  }

  @Test
  fun `diff empty file`() {
    // given
    val newEmptyFileDiff = """
      diff --git a/file-b.txt b/file-b.txt
      new file mode 100644
      index 0000000..e69de29
      --- /dev/null
      +++ b/file-b.txt
      
    """.trimIndent()

    // when
    val diffHtml = FormattedDiff.from(newEmptyFileDiff).toHtml()

    // then
    Approvals.verify(diffHtml)
  }

  @Test
  fun `escape HTML characters`() {
    // given
    val rawXmlDiff = """
      diff --git a/app/src/main/res/drawable/background_facility_progress_shimmer.xml b/app/src/main/res/drawable/background_facility_progress_shimmer.xml
      new file mode 100644
      index 0000000..0a87cfd
      --- /dev/null
      +++ b/app/src/main/res/drawable/background_facility_progress_shimmer.xml
      @@ -0,0 +1,5 @@
      +<?xml version="1.0" encoding="utf-8"?>
      +<shape xmlns:android="http://schemas.android.com/apk/res/android">
      +  <corners android:radius="2dp" />
      +  <solid android:color="#EBEFF0" />
      +</shape>
      
    """.trimIndent()

    // when
    val diffHtml = FormattedDiff.from(rawXmlDiff).toHtml()

    // then
    Approvals.verify(diffHtml)
  }

  @Test
  fun `insertion padding space`() {
    // given
    val insertionRawDiff = """
      diff --git a/.codeclimate.yml b/.codeclimate.yml
      index 4a85e39..62c22dc 100644
      --- a/.codeclimate.yml
      +++ b/.codeclimate.yml
      @@ -46,3 +46,4 @@
         - "**/src/main/java/org/simple/clinic/storage/Migration_*.kt"
         - "**/build/"
         - "**/src/main/java/org/simple/clinic/widgets/OmegaCenterIconButton.java"
      +  - "router/"
      
    """.trimIndent()

    // when
    val diffHtml = FormattedDiff.from(insertionRawDiff).toHtml()

    // then
    Approvals.verify(diffHtml)
  }
}
