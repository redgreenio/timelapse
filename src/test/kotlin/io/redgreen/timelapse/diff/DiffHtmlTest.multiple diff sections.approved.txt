<html>
  <head>
    <style>
      #page {
        padding: 0;
        margin: 0;
      }

      table {
        width: 100%;
        border-collapse: collapse;
      }

      table td {
        font-family: "Fira Code", "JetBrains Mono", "Lucida Console";
        font-size: small;
      }

      table tr td:first-child {
        padding-left: 24px;
      }

      .deletion {
        background-color: #ffeef0;
      }

      .insertion {
        background-color: #e6ffed;
      }

      .blob {
        padding-left: 10px;
      }

      .diff-section {
        background-color: #f1f8ff;
      }

      .line-number {
        width:1%;
        white-space:nowrap;
        text-align: right;
        vertical-align: top;
        padding: 2px 8px 2px 8px;
        color: rgba(27, 31, 35, 0.4);
      }

      .deletion .line-number {
        background-color: #ffdce0;
      }

      .insertion .line-number {
        background-color: #bef5cb;
      }

      .unmodified .line-number {
        background-color: rgba(230, 230, 230, 0.1);
      }

      .diff-section .line-number {
        background-color: #dbedff;
      }
    </style>
  </head>
  <body id="page">
    <table>
      <tbody>
        <tr class="diff-section">
          <td class="line-number">​</td><td class="line-number">​</td><td class="blob"></td>
        </tr>
        <tr class="unmodified">
          <td class="line-number">1</td><td class="line-number">1</td><td class="blob">&nbsp;&nbsp;apply plugin: 'com.android.application'</td>
        </tr>
        <tr class="unmodified">
          <td class="line-number">2</td><td class="line-number">2</td><td class="blob">&nbsp;&nbsp;apply plugin: 'kotlin-android'</td>
        </tr>
        <tr class="unmodified">
          <td class="line-number">3</td><td class="line-number">3</td><td class="blob">&nbsp;&nbsp;apply plugin: 'kotlin-android-extensions'</td>
        </tr>
        <tr class="insertion">
          <td class="line-number"></td><td class="line-number">4</td><td class="blob">+&nbsp;apply plugin: 'kotlin-kapt'</td>
        </tr>
        <tr class="unmodified">
          <td class="line-number">4</td><td class="line-number">5</td><td class="blob">&nbsp;&nbsp;</td>
        </tr>
        <tr class="unmodified">
          <td class="line-number">5</td><td class="line-number">6</td><td class="blob">&nbsp;&nbsp;android {</td>
        </tr>
        <tr class="unmodified">
          <td class="line-number">6</td><td class="line-number">7</td><td class="blob">&nbsp;&nbsp;&nbsp;&nbsp;compileSdkVersion versions.compileSdk</td>
        </tr>
        <tr class="diff-section">
          <td class="line-number">​</td><td class="line-number">​</td><td class="blob"></td>
        </tr>
        <tr class="unmodified">
          <td class="line-number">28</td><td class="line-number">29</td><td class="blob">&nbsp;&nbsp;&nbsp;&nbsp;implementation &quot;com.android.support:recyclerview-v7:$versions.supportLib&quot;</td>
        </tr>
        <tr class="unmodified">
          <td class="line-number">29</td><td class="line-number">30</td><td class="blob">&nbsp;&nbsp;</td>
        </tr>
        <tr class="unmodified">
          <td class="line-number">30</td><td class="line-number">31</td><td class="blob">&nbsp;&nbsp;&nbsp;&nbsp;implementation &quot;com.jakewharton.timber:timber:$versions.timber&quot;</td>
        </tr>
        <tr class="insertion">
          <td class="line-number"></td><td class="line-number">32</td><td class="blob">+&nbsp;&nbsp;&nbsp;implementation &quot;com.google.dagger:dagger:$versions.dagger&quot;</td>
        </tr>
        <tr class="insertion">
          <td class="line-number"></td><td class="line-number">33</td><td class="blob">+&nbsp;&nbsp;&nbsp;kapt &quot;com.google.dagger:dagger-compiler:$versions.dagger&quot;</td>
        </tr>
        <tr class="unmodified">
          <td class="line-number">31</td><td class="line-number">34</td><td class="blob">&nbsp;&nbsp;}</td>
        </tr>
        <tr class="unmodified">
          <td class="line-number">32</td><td class="line-number">35</td><td class="blob">&nbsp;&nbsp;</td>
        </tr>
      </tbody>
    </table>
  </body>
</html>
